File: Algorithms/p354_Russian_Doll_Envelopes/p354_Russian_Doll_Envelopes.py
class Solution(object):
    def maxEnvelopes(self, envelopes):
        """
        :type envelopes: List[List[int]]
        :rtype: int
        """
        envelopes.sort(key=lambda e: (e[0],-e[1]))
        # min_h[i]: min h of i-length sequence's last envelope so far
        min_h = [0]*(len(envelopes)+1)
        max_len = 0
        for e in envelopes:
            j = bisect.bisect_left(min_h,e[1],0,max_len+1)
            min_h[j] = e[1]
            max_len = max(max_len,j)
        return max_len

File: Algorithms/p354_Russian_Doll_Envelopes/p354_Russian_Doll_Envelopes.cpp
class Solution {
public:
    int maxEnvelopes(vector<pair<int, int>>& envelopes) {
        sort(envelopes.begin(),envelopes.end(),[](pair<int,int>& e1, pair<int,int>& e2){ return e1.first<e2.first||e1.first==e2.first&&e1.second>e2.second; });
        // min_h[i]: min h of i-length sequence's last envelope so far
        vector<int> min_h(envelopes.size()+1,0);
        int max_len = 0;
        for(auto &e : envelopes){
            int j = lower_bound(min_h.begin(),min_h.begin()+(max_len+1),e.second)-min_h.begin();
            min_h[j] = e.second;
            if(j>max_len) max_len = j;
        }
        return max_len;
    }
};
